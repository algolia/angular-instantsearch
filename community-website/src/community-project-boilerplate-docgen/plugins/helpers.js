const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
