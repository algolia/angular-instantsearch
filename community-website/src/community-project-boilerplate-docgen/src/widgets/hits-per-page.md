---
title: Widgets - Hits Per Page
layout: widget.pug
---

# Hits per page

## Description

The Hits per page widget displays a dropdown menu to let the user change the number of displayed hits.

## Options

| Attribute      | Type                                                  | Description
| -              | -                                                     | -
| `items`        | `{value: number, label: string, default?: boolean}[]` | List of available options
| `autoHideContainer?` | `boolean`  | Hides the results per pages if there's no item to display

## Code example

```html
<ais-hits-per-page
  [items]="[
    { value: 3, label: '3 per page' },
    { value: 5, label: '5 per page' },
    { value: 10, label: '10 per page' }
  ]"
>
</ais-hits-per-page>
```
