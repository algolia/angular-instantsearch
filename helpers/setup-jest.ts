import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
