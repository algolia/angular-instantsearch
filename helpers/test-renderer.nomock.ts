import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
