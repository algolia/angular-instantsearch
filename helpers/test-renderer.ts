import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
