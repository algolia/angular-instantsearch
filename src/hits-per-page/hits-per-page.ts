import { Component, Input, Inject, forwardRef } from '@angular/core';

import { connectHitsPerPage } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
import { noop } from '../utils';

export type ResultsPerPageState = {
  items: {}[];
  refine: Function;
};

@Component({
  selector: 'ais-hits-per-page',
  template: `
    <div
      [class]="cx()"
      *ngIf="!isHidden"
    >
      <select
        [class]="cx('select')"
        (change)="state.refine($event.target.value)"
      >
        <option
          [class]="cx('option')"
          *ngFor="let item of state.items"
          [value]="item.value"
          [selected]="item.isRefined"
        >
          {{item.label}}
        </option>
      </select>
    </div>
  `,
})
export class NgAisHitsPerPage extends BaseWidget {
  @Input()
  public items: {
    value: number;
    label: string;
    default?: boolean;
  }[];

  public state: ResultsPerPageState = {
    items: [],
    refine: noop,
  };

  get isHidden() {
    return this.state.items.length === 0 && this.autoHideContainer;
  }

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('HitsPerPage');
  }

  public ngOnInit() {
    this.createWidget(connectHitsPerPage, { items: this.items });
    super.ngOnInit();
  }
}
