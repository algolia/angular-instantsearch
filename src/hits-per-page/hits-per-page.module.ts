import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisHitsPerPage } from './hits-per-page';

@NgModule({
  declarations: [NgAisHitsPerPage],
  entryComponents: [NgAisHitsPerPage],
  exports: [NgAisHitsPerPage],
  imports: [CommonModule],
})
export class NgAisHitsPerPageModule {}
