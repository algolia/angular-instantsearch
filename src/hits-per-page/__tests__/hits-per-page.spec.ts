import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisHitsPerPage } from "../hits-per-page";

const render = createRenderer({
  defaultState: {
    items: [
      { value: 10, label: "10 per page", isRefined: true },
      { value: 20, label: "20 per page" },
      { value: 30, label: "30 per page" }
    ],
    refine: jest.fn()
  },
  template: "<ais-hits-per-page></ais-hits-per-page>",
  TestedWidget: NgAisHitsPerPage
});

describe("HitsPerPage", () => {
  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should refine() when selecting another option", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const select = fixture.debugElement.nativeElement.querySelector("select");
    select.value = "20";
    select.dispatchEvent(new Event("change"));

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith("20");
  });

  it("should be hidden with autoHideContainer", () => {
    const fixture = render();

    fixture.componentInstance.testedWidget.autoHideContainer = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
