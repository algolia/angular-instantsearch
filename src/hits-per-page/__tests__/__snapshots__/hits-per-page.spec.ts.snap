// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HitsPerPage renders markup with state 1`] = `
<ng-component
  testedWidget={[Function NgAisHitsPerPage]}
>
  
        <ais-instantsearch>
    
          
    <ais-hits-per-page>
      
      
      
      <div
        class="ais-HitsPerPage"
      >
        
        
        <select
          class="ais-HitsPerPage-select"
        >
          
          
          
          <option
            class="ais-HitsPerPage-option"
            value="10"
          >
            
            10 per page
          
          </option>
          <option
            class="ais-HitsPerPage-option"
            value="20"
          >
            
            20 per page
          
          </option>
          <option
            class="ais-HitsPerPage-option"
            value="30"
          >
            
            30 per page
          
          </option>
          
        
        </select>
        
      
      </div>
      
    
    </ais-hits-per-page>
    
        
  </ais-instantsearch>
      
</ng-component>
`;

exports[`HitsPerPage renders markup without state 1`] = `
<ng-component
  testedWidget={[Function NgAisHitsPerPage]}
>
  
        <ais-instantsearch>
    
          
    <ais-hits-per-page>
      
      
      
      <div
        class="ais-HitsPerPage"
      >
        
        
        <select
          class="ais-HitsPerPage-select"
        >
          
          
          
          
        
        </select>
        
      
      </div>
      
    
    </ais-hits-per-page>
    
        
  </ais-instantsearch>
      
</ng-component>
`;

exports[`HitsPerPage should be hidden with autoHideContainer 1`] = `
<ng-component
  testedWidget={[Function NgAisHitsPerPage]}
>
  
        <ais-instantsearch>
    
          
    <ais-hits-per-page>
      
      
      
      
    
    </ais-hits-per-page>
    
        
  </ais-instantsearch>
      
</ng-component>
`;
